src/random.rs

```diff
@@ -0,0 +1,325 @@
+    /// Signature RandomAlgorithmId).
+    ///
+    /// Standard: FIPS 186-2.
+    ///
+    /// *Windows 8*: Support for FIPS 186-3 begins.
+    FIPS186DSARNG,
+}
+
+impl Default for RandomAlgorithm {
+    fn default() -> Self {
+        RandomAlgorithm::SystemPreferred
+    }
+}
+
+impl<'a> TryFrom<&'a str> for RandomAlgorithmId {
+    type Error = &'a str;
+
+    fn try_from(value: &'a str) -> Result<RandomAlgorithmId, Self::Error> {
+        match value {
+            BCRYPT_RNG_ALGORITHM => Ok(RandomAlgorithmId::RNG),
+            BCRYPT_RNG_DUAL_EC_ALGORITHM => Ok(RandomAlgorithmId::DUALECRNG),
+            BCRYPT_RNG_FIPS186_DSA_ALGORITHM => Ok(RandomAlgorithmId::FIPS186DSARNG),
+            _ => Err(value),
+        }
+    }
+}
+
+impl Into<&'static str> for RandomAlgorithmId {
+    fn into(self) -> &'static str {
+        match self {
+            RandomAlgorithmId::RNG => BCRYPT_RNG_ALGORITHM,
+            RandomAlgorithmId::DUALECRNG => BCRYPT_RNG_DUAL_EC_ALGORITHM,
+            RandomAlgorithmId::FIPS186DSARNG => BCRYPT_RNG_FIPS186_DSA_ALGORITHM,
+        }
+    }
+}
+
+impl fmt::Display for RandomAlgorithmId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", Into::<&'static str>::into(*self))
+    }
+}
+
+impl From<RandomAlgorithmId> for RandomAlgoHandle {
+    fn from(value: RandomAlgorithmId) -> RandomAlgoHandle {
+        let id: &str = value.into();
+
+        AlgoHandle::open(id)
+            .map(RandomAlgoHandle)
+            .expect("Couldn't open random algorithm provider")
+    }
+}
+
+impl From<RandomAlgoHandle> for RandomAlgorithm {
+    fn from(value: RandomAlgoHandle) -> RandomAlgorithm {
+        RandomAlgorithm::Specified(value)
+    }
+}
+
+impl From<RandomAlgoHandle> for AlgoHandle {
+    fn from(value: RandomAlgoHandle) -> AlgoHandle {
+        value.0
+    }
+}
+
+impl From<RandomAlgorithmId> for RandomAlgorithmKind {
+    fn from(value: RandomAlgorithmId) -> RandomAlgorithmKind {
+        RandomAlgorithmKind::Specified(value)
+    }
+}
+
+impl RandomAlgorithmKind {
+    /// Opens a random algorithm provider capable of generating random numbers.
+    pub fn open(self) -> RandomAlgorithm {
+        match self {
+            Self::SystemPreferred => RandomAlgorithm::SystemPreferred,
+            Self::Specified(kind) => RandomAlgorithm::Specified(kind.into()),
+        }
+    }
+}
+
+impl RandomAlgorithmId {
+    /// Opens a random algorithm provider capable of generating random numbers.
+    pub fn open(self) -> RandomAlgorithm {
+        RandomAlgorithmKind::from(self).open()
+    }
+}
+
+impl RandomAlgorithm {
+    /// Fills a buffer with random bytes.
+    pub fn gen_random(&self, buffer: &mut [u8]) -> crate::Result<()> {
+        Self::gen_random_with_opts(self, buffer, RandomOptions::default())
+    }
+
+    /// Fills a buffer with random bytes. Specify `opts` for additional options.
+    pub fn gen_random_with_opts(
```

Comment:
While I agree that this would simplify things and that this should probably be not used as you're saying, I don't think we get to be opinionated here as just a type-safe wrapper API.

After all, what is valid to do on the lower level should be possible to do here as well.

Reply:
I mean `BCRYPT_USE_SYSTEM_PREFERRED_RNG` will be usable through the `RandomAlgorithmId::SystemPrefered`, so the only thing left is the `BCRYPT_RNG_USE_ENTROPY_IN_BUFFER` which is not supported on Windows 7 and lower (extended support for Windows 7 ends in 10 days).

I guess we could simply add a `gen_random_with_entropy_in_buffer` method and mark it as deprecated, but tbh, I wouldn't mind just not supporting it.